From ecf9948014d367b6fdf8d2ac9aa23f029ab872ff Mon Sep 17 00:00:00 2001
From: Iris Chang <iris.chang@mediatek.com>
Date: Tue, 25 Sep 2018 13:46:11 +0800
Subject: [PATCH 08/10] APP may display abnormally in landscape LCM

When PRIMARY_DISPLAY_ORIENTATION is 90, 180 or 270, there are two issues as below.
1. APP preview displays in wrong direction
2. When APP is launched, APP may display in landscape mode firstly and then change
to portrait mode

The root cause is WMS/AMS use captureLayers instead of captureScreen, the captured
layer displays in wrong direction with surfaceflinger correction.

As captureLayers's CaptureFill is CLEAR, captureScreen's CaptureFill is default
OPAQUE, we can use ReaderArea CaptureFill to separate captureLayers and
captureScreen. And SurfaceFlinger disables the correction if cpatureLayers has been
called.

Test: 1. APP preview displays in correct direction
2. APP is launched normally
3. The  direction of captrueScreen is correct

Change-Id: I759448404970202a88c52b2a21549a777698fb0b
---
 services/surfaceflinger/SurfaceFlinger.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index 41433cb3d..942441058 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -5062,7 +5062,8 @@ void SurfaceFlinger::renderScreenImplLocked(const RenderArea& renderArea,
     if (sourceCrop.width() == 0 || sourceCrop.height() == 0 || !sourceCrop.isValid()) {
         sourceCrop.setLeftTop(Point(0, 0));
         sourceCrop.setRightBottom(Point(raWidth, raHeight));
-    } else if (mPrimaryDisplayOrientation != DisplayState::eOrientationDefault) {
+    } else if (mPrimaryDisplayOrientation != DisplayState::eOrientationDefault &&
+               renderArea.getCaptureFill() != RenderArea::CaptureFill::CLEAR) {
         Transform tr;
         uint32_t flags = 0x00;
         switch (mPrimaryDisplayOrientation) {
@@ -5102,7 +5103,8 @@ void SurfaceFlinger::renderScreenImplLocked(const RenderArea& renderArea,
     engine.checkErrors();
 
     Transform::orientation_flags rotation = renderArea.getRotationFlags();
-    if (mPrimaryDisplayOrientation != DisplayState::eOrientationDefault) {
+    if (mPrimaryDisplayOrientation != DisplayState::eOrientationDefault &&
+        renderArea.getCaptureFill() != RenderArea::CaptureFill::CLEAR) {
         // convert hw orientation into flag presentation
         // here inverse transform needed
         uint8_t hw_rot_90  = 0x00;
-- 
2.17.1

